### PR TITLE
Create AP_HAL::Util::internal_errors(...)

### PR DESCRIPTION
**Pull request by peterbarker**
To allow notification of  "can't happen" things happening.

### CODE REVIEW

File: libraries/GCS_MAVLink/GCS.h [hunk 1]
```diff
@@ -772,6 +772,9 @@ class GCS
     // update uart pass-thru
     void update_passthru();
 
+    static uint16_t sys_status_errors1();
+    static uint16_t sys_status_errors2();
```
Comment (OXINARF) [inline libraries/GCS_MAVLink/GCS.h]: Unused?

**Comment by magicrub** [comment]
This may be a good candidate to use on PR https://github.com/ArduPilot/ardupilot/pull/10483

**Comment by peterbarker** [comment]
@OXINARF There are plenty of places in the code that look like this:

```
#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
        AP_HAL::panic("next_deferred_bucket_message_to_send called on empty bucket");
#endif
```

and 

```
            #if CONFIG_HAL_BOARD == HAL_BOARD_SITL
                AP_HAL::panic("Mission checks requested, but no rally was allocated");
            #endif // CONFIG_HAL_BOARD == HAL_BOARD_SITL
```

The problem is that on real boards we currently fail silently - we never log those errors or (most of the time) send statustexts.

The concept here is to create a function which gives us that feedback - in logs and live, over mavlink.  Any GCS should be able to monitor those bits - and if one goes on carry on like a flaming galah.

`AP_Logger` was only used as an example here.  I should have added another library for clarity - but this is the sort of thing that bitrots...

I think I may have gone overboard on the bits used from `AP_Logger`, mind you.  However, I was thinking the bits would be volatile on a version-to-version basis (at odds, I know, with the MAVLink thing, but all the GCS really needs to know is "any bit set is very, very bad").  If they are "volatile" then burning them up doesn't matter too much here.

**Comment by OXINARF** [comment]
@peterbarker Ok, now I understand the purpose (although I'm not sure if it is really relevant to send individual bits over MAVLink), but can't we put that in AP_Common or a different/new library? The HAL should really be as small as possible, we've been removing stuff from it into separate libraries, this seems to go in the opposite direction.

**Comment by peterbarker** [comment]
On Fri, 22 Feb 2019, Francisco Ferreira wrote:

> 
> @peterbarker Ok, now I understand the purpose (although I'm not sure if it is really relevant to send individual bits over MAVLink), but can't we put that
> in AP_Common or a different/new library? The HAL should really be as small as possible, we've been removing stuff from it into separate libraries, this
> seems to go in the opposite direction.

Seems a reasonable request.

hal was convenient because it's ubiquitous, as this code needs to be. 
Also, it let me easily change behaviour if we were in SITL without relying 
on #ifdefs.

Past this I'd probably have to move to a static mask of what errors have 
occured somewhere, and possible having AP::internal_error(...).

You think perhaps AP_InternalError?!  Seems a bit overkill :-)

AP_Common - not sure about that either...

Hmmm.

**Comment by OXINARF** [comment]
> hal was convenient because it's ubiquitous

All libraries that have a singleton returning a reference are too. That would be the case with this new library as well.

> Past this I'd probably have to move to a static mask of what errors have occured somewhere, and possible having AP::internal_error(...).
>
> You think perhaps AP_InternalError?! Seems a bit overkill :-)

Seems perfect to me 😜

File: libraries/GCS_MAVLink/GCS.h (hunk 1 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS.h]: Removed.  I added it as a hint as to what was going on.

**Pull request merged**